### ISSUE TITLE
Use go-cmp instead of reflect.DeepEqual

### ISSUE THREAD

**Issue by ernesto-jimenez** [stretchr/testify#535]
The package `github.com/google/go-cmp` is better suited for comparison in tests than `reflect.DeepEqual`.

We should replace our `DeepEqual` usage with `go-cmp`.

Replacement should be backwards-compatible.

**Comment by dnephin** [comment]
From what I've seen having used [go-cmp to create an assertion](https://github.com/gotestyourself/gotestyourself/pull/34) there's no way to make it backwards compatible. 

`go-cmp` requires that for each struct in a comparison that has any unexported fields you must explicitly use either `AllowUnexporterted` or `IgnoreUnexported`. This requirement seems like one of the advantages of using `go-cmp` instead of reflect.DeepEqual.

**Comment by runner-mei** [comment]
advantages of using go-cmp instead of reflect.DeepEqual.

````patch
                        Diff:
                        --- Expected
                        +++ Actual
                        @@ -12,3 +12,3 @@
                           CurrentValue: (string) (len=2) "ww",
                        -  TriggeredAt: (time.Time) 2018-02-07 08:54:01.123 +0000 UTC,
                        +  TriggeredAt: (time.Time) 2018-02-07 16:54:01.123 +0800 HKT,
                           ConfirmContent: (string) "",
````

**Comment by mattayes** [comment]
@dsnet might be a good person to loop into this conversation (if they're interested).

**Comment by posener** [comment]
Tried to implement this.
Used what @dsnet suggested with the `deepAllowUnexported` option.
The go-cmp got into an infinite recursion in the testify tests. I guess because of a cyclic list.
I opened an issue in go-cmp, but it seems like it is not going to be supported.
https://github.com/google/go-cmp/issues/74

**Comment by mattayes** [comment]
@posener Do you have that branch around still?

**Comment by posener** [comment]
can push it, nothing to fancy
https://github.com/posener/testify/tree/cmp

**Comment by posener** [comment]
If [this](https://github.com/google/go-cmp/pull/78) is merged, we can have `go-cmp` used here

**Comment by ernesto-jimenez** [comment]
@posener being able to use `go-cmp` would be fantastic. 🙌

**Comment by posener** [comment]
@ernesto-jimenez another issue here: I guess we are intending also to use the `cmp.Diff` function. However, it's output is different then the current one. I guess this is API breaking change. Any ideas?

**Comment by posener** [comment]
I've used my branch in [go-cmp](https://github.com/posener/go-cmp/tree/detect-cycles2) and my branch [testify](https://github.com/posener/testify/tree/cmp).

* It now runs (no stack overflows...)
* Diff is different since go-fmt has different format.
* Types equality fails for some reason.

Here are the result of the tetify unit-tests: [results.txt](https://github.com/stretchr/testify/files/1827183/results.txt)

**Comment by dnephin** [comment]
How are you planning on accepting https://godoc.org/github.com/google/go-cmp/cmp#Option to `Equal` ? Without `Option`, and always using `deepAllowUnexported`, the only advantage over `reflect.DeepEqual` is the change in output format.

**Comment by posener** [comment]
Another blocking issue for using go-cmp: https://github.com/google/go-cmp/issues/80.

**Comment by posener** [comment]
@dnephin , I think the main advantage is that it gives more control over comparing: For example, if you want extra control on comparing time objects: ignore timezone or wall time, it could be done in the `cmp.Equal` but not in `reflect.DeepEqual`.

**Comment by dsnet** [comment]
To expand: `reflect.DeepEqual` does a decent job at comparisons and has serve the Go community well for many years. However, compatibility issues due to its pervasive usage has increased as the scale of Go code also increases. For example, when monotonic timestamps were added, it could not compare `time.Time` well as it would implicitly compare upon the unexported monotonic timestamps fields. It had no understanding of the `time.Time.Equal` method that should probably be used instead.

`cmp.Equal` tries to do a better job at this by:
1) Allowing the user to customize any portion of the comparison.
2) Allowing the author of a type to specify the correct definition of comparison (by searching for an `Equal` method).
3) Falling back on `reflect.DeepEqual`-like behavior.

I should note that `cmp` does not claim to be the perfect comparison 100% of the time either, but the ability to specify `Options` allows an easy fix in the rare event of a comparison failure. For example, a type that your package depends on may add a new field that causes your comparison to fail. In some cases you would have wanted to ignore any newly added fields. In other cases, you do want to compare across newly added fields as well. Whatever decision the comparison library chooses to do will never be correct all the time, and Options help alleviate these rare (but inevitable) cases.

**Comment by dnephin** [comment]
@posener Yes I understand the value of `go-cmp`. I have been using it instead of `reflect.DeepEqual`.

My question was:
> How are you planning on accepting https://godoc.org/github.com/google/go-cmp/cmp#Option to `assert.Equal` ? 

`testify/assert.Equal` already has a var args, so you can't use a var args like `cmp.Equal` (and [as I have](https://godoc.org/github.com/gotestyourself/gotestyourself/assert#DeepEqual)).

If you only apply a static set of `Options` that mimic the behaviour of `reflect.DeepEqual` you are not actually gaining any of the value of having `Options`.

So either there needs to be a different function added, or you break compatiblity and add an Options arg. Or is there some other option I'm missing?

**Comment by posener** [comment]
I agree - and we are also breaking API by changing the diff format.

**Comment by posener** [comment]
Rethinking:

I see two more options:
1. We can just use cmp.Equal without any user-define problems, and define "general" options that can solve issues in testify, such as time comparing.

2. We can have an ugly workaround for user define options:
  * We can pass compare options to [assert.New](https://github.com/stretchr/testify/blob/c7668ea997d68b07ceaa34c46cef1419c39a2374/assert/forward_assertions.go#L10)
  * Then pass an object that wraps testing.T to the Equal function [here](https://github.com/stretchr/testify/blob/master/assert/assertion_forward.go#L129)
  * Then have type assertion in the Equal function, and extract the options there...

**Comment by icholy** [comment]
Another option is to allow overriding default behaviour.

``` gotype EqFunc func (a, b interface{}) bool

t.Eq = func(a, b interface{}) bool {
  return reflect.DeepEqual(a, b)
}
```

**Comment by ansel1** [comment]
Yet another option: leave the current function as is, to maintain compatibility, and add a new function which uses go-cmp (and takes options, etc).  EqualEx or Equivalent or Comparable or something.

**Comment by shaunc** [comment]
This would also allow fixing issues such as: 
https://github.com/stretchr/testify/issues/624

I have NaNs in structures, and want a comparison that I can use to assert equality, which I can do with go-cmp.

**Comment by e11ni** [comment]
I have similar problem with floats, in my structures there are floats that cannot be compared just for `==`. Go-cmp provide approximate equality for floats.

> Yet another option: leave the current function as is, to maintain compatibility, and add a new function which uses go-cmp (and takes options, etc). EqualEx or Equivalent or Comparable or something.

What about this option? I can contribute to make it happen.

**Comment by mbana** [comment]
Any news on this?

**Comment by lcd1232** [comment]
Any updates?

**Comment by allenhaozi** [comment]
> The package `github.com/google/go-cmp` is better suited for comparison in tests than `reflect.DeepEqual`.
> 
> We should replace our `DeepEqual` usage with `go-cmp`.
> 
> Replacement should be backwards-compatible.

found a bad case 
`type Item struct {
	Media string
	Size  int64
}

func main() {

	standard := []Item{
		Item{
			Media: "SSD",
			Size:  100,
		},

		Item{
			Media: "HDD",
			Size:  200,
		},
	}

	test1 := []Item{
		Item{
			Media: "HDD",
			Size:  200,
		},
		Item{
			Media: "SSD",
			Size:  100,
		},
	}

	fmt.Println(cmp.Equal(standard, test1))

}`

want : true , got: fase

**Comment by posener** [comment]
I guess that is ok.
We can pass default options that will imitate the behavior of `assert.Equal`.

See https://github.com/stretchr/testify/pull/579/files

**Comment by cbandy** [comment]
@allenhaozi I think you are trying or want to assert with `ElementsMatch`.

**Comment by boyan-soubachov** [comment]
Of all the options, I think it would make sense to gauge community interest in making a `testify/v2` that's based on `go-cmp`. That way we can see follow usage patterns. Unfortunately that also means two versions of the repo to maintain but we can use the data provided to us and start deprecating the lesser-used version where necessary over time.

I don't think we can (feasibly) get around the diff output changes without breaking the API and the point about `go-cmp`'s power being in its options (thanks @posener) make too much sense to not expose them to our end users.

**Comment by posener** [comment]
The new go-cmp [Reporter option](https://pkg.go.dev/github.com/google/go-cmp/cmp?tab=doc#Reporter) can be used to create custom diff.

**Comment by rustyx** [comment]
There is already an assertion framework based on `go-cmp`, [`gotest.tools/assert`](https://pkg.go.dev/gotest.tools/assert)

So a project like `testify/v2` might be too late, unfortunately.

**Comment by yangxikun** [comment]
When I need advance comparison, I use `a.Empty(cmp.Diff(x, y, opts...))` to compose go-cmp with testify.
